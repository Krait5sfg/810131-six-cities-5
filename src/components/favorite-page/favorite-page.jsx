import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import FavoriteLocationList from '../favorite-location-list/favorite-location-list';
import {OfferPropTypes} from '../../utils/property-type';
import User from '../user/user';
import {connect} from 'react-redux';
import {getFavoriteOffers} from '../../store/api-actions';
import Header from '../header/header';

class FavoritePage extends PureComponent {

  componentDidMount() {
    const {updateFavoriteOffers} = this.props;
    updateFavoriteOffers();
  }

  render() {
    const {favoriteOffers, onLinkEmailClick} = this.props;
    const isEmpty = !favoriteOffers.length;
    const classNameForDiv = isEmpty ? `page--favorites-empty` : ``;
    const classNameForMain = isEmpty ? `page__main--favorites-empty` : ``;
    const classNameForFavoriteSection = isEmpty ? `favorites--empty` : ``;

    const allCities = favoriteOffers.map((offer) => offer.city);
    const uniqueCities = [...new Set(allCities)];

    const favoriteLocationListElements = uniqueCities.map((city, index) => {
      const cityFavoriteOffers = favoriteOffers.filter((offer) => offer.city === city);
      return (
        <FavoriteLocationList
          city={city}
          key={index}
          cityFavoriteOffers={cityFavoriteOffers} />
      );
    });

    return (
      <div className={`page ${classNameForDiv}`}>
        <Header>
          <User onLinkEmailClick={onLinkEmailClick} />
        </Header>

        <main className={`page__main page__main--favorites ${classNameForMain}`}>
          <div className="page__favorites-container container">
            <section className={`favorites ${classNameForFavoriteSection}`}>
              {isEmpty ?
                <React.Fragment>
                  <h1 className="visually-hidden">Favorites (empty)</h1>
                  <div className="favorites__status-wrapper">
                    <b className="favorites__status">Nothing yet saved.</b>
                    <p className="favorites__status-description">Save properties to narrow down search or plan yor future trips.</p>
                  </div>
                </React.Fragment>
                :
                <React.Fragment>
                  <h1 className="favorites__title">Saved listing</h1>
                  <ul className="favorites__list">
                    {favoriteLocationListElements}
                  </ul>
                </React.Fragment>
              }
            </section>

          </div>
        </main>
        <footer className="footer container">
          <a className="footer__logo-link" href="/">
            <img className="footer__logo" src="img/logo.svg" alt="6 cities logo" width="64" height="33" />
          </a>
        </footer>
      </div >
    );
  }
}

FavoritePage.propTypes = {
  favoriteOffers: PropTypes.arrayOf(OfferPropTypes).isRequired,
  onLinkEmailClick: PropTypes.func.isRequired,
  updateFavoriteOffers: PropTypes.func.isRequired
};

const mapStateToProps = (({DATA}) => {
  return {
    favoriteOffers: DATA.favoriteOffers
  };
});

const mapDispatchToProps = ((dispatch) => ({
  updateFavoriteOffers() {
    dispatch(getFavoriteOffers());
  },
}));

export {FavoritePage};
export default connect(mapStateToProps, mapDispatchToProps)(FavoritePage);
