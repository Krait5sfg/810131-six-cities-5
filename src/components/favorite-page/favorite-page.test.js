import React from 'react';
import renderer from 'react-test-renderer';
import {FavoritePage} from './favorite-page';
import {BrowserRouter} from 'react-router-dom';

const offers = [
  {
    city: ``,
    accommodation: {
      bedroomsCount: 0,
      features: [],
      guestsLimit: 0,
      isPremium: true,
      price: 0,
      rating: 0,
      title: ``,
      type: ``,
    },
    cityLocation: {latitude: 0, longitude: 0, zoom: 0},
    coordinates: [0, 0],
    description: ``,
    host: {
      avatar: ``,
      isPro: true,
      name: ``,
    },
    id: 1,
    images: [],
    isFavorite: true,
    previewImage: ``,
  },
  {
    city: ``,
    accommodation: {
      bedroomsCount: 0,
      features: [],
      guestsLimit: 0,
      isPremium: true,
      price: 0,
      rating: 0,
      title: ``,
      type: ``,
    },
    cityLocation: {latitude: 0, longitude: 0, zoom: 0},
    coordinates: [0, 0],
    description: ``,
    host: {
      avatar: ``,
      isPro: true,
      name: ``,
    },
    id: 2,
    images: [],
    isFavorite: true,
    previewImage: ``,
  }
];

jest.mock(`../header/header`, () => `Header`);
jest.mock(`../user/user`, () => `User`);
jest.mock(`../favorite-location-list/favorite-location-list`, () => `FavoriteLocationList`);

describe(`FavoritePage render correctly`, () => {
  it(`FavoritePage with favorite offers`, () => {
    const tree = renderer.create(<BrowserRouter><FavoritePage
      favoriteOffers={offers}
      onLinkEmailClick={() => {}}
      updateFavoriteOffers={() => {}}
    /></BrowserRouter>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`FavoritePage without favorite offers`, () => {
    const tree = renderer.create(<BrowserRouter><FavoritePage
      favoriteOffers={[]}
      onLinkEmailClick={() => {}}
      updateFavoriteOffers={() => {}}
    /></BrowserRouter>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
