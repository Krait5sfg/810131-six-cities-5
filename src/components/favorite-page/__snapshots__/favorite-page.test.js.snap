// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritePage render correctly FavoritePage with favorite offers 1`] = `
<div
  className="page "
>
  <Header>
    <User
      onLinkEmailClick={[Function]}
    />
  </Header>
  <main
    className="page__main page__main--favorites "
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites "
      >
        <h1
          className="favorites__title"
        >
          Saved listing
        </h1>
        <ul
          className="favorites__list"
        >
          <FavoriteLocationList
            cityFavoriteOffers={
              Array [
                Object {
                  "accommodation": Object {
                    "bedroomsCount": 1,
                    "features": Array [
                      "Dishwasher",
                      "Towels",
                      "Washer",
                      "Fridge",
                      "Breakfast",
                      "Laptop friendly workspace",
                      "Baby seat",
                      "Air conditioning",
                    ],
                    "guestsLimit": 2,
                    "isPremium": true,
                    "price": 241,
                    "rating": 4.4,
                    "title": "The Joshua Tree House",
                    "type": "room",
                  },
                  "cityLocation": Object {
                    "latitude": 51.225402,
                    "longitude": 6.776314,
                    "zoom": 13,
                  },
                  "coordinates": Array [
                    51.237402,
                    6.797314,
                  ],
                  "description": "This is a place for dreamers to reset, reflect, and create. Designed with a 'slow' pace in mind, our hope is that you enjoy ",
                  "host": Object {
                    "avatar": "img/avatar-angelina.jpg",
                    "isPro": true,
                    "name": "Angelina",
                  },
                  "id": 1,
                  "images": Array [
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/12.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/1.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg",
                  ],
                  "isFavorite": false,
                  "previewImage": "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/6.jpg",
                },
                Object {
                  "accommodation": Object {
                    "bedroomsCount": 2,
                    "features": Array [
                      "Laptop friendly workspace",
                      "Baby seat",
                      "Air conditioning",
                    ],
                    "guestsLimit": 4,
                    "isPremium": false,
                    "price": 300,
                    "rating": 4.5,
                    "title": "The Joshua Tree House",
                    "type": "hotel",
                  },
                  "cityLocation": Object {
                    "latitude": 51.225402,
                    "longitude": 6.776314,
                    "zoom": 13,
                  },
                  "coordinates": Array [
                    51.247402,
                    6.777314,
                  ],
                  "description": "This is a place for dreamers to reset, reflect, and create. Designed with a 'slow' pace in mind, our hope is that you ",
                  "host": Object {
                    "avatar": "img/avatar-angelina.jpg",
                    "isPro": true,
                    "name": "Angelina",
                  },
                  "id": 2,
                  "images": Array [
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/12.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/1.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/15.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/2.jpg",
                    "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/5.jpg",
                  ],
                  "isFavorite": true,
                  "previewImage": "https://assets.htmlacademy.ru/intensives/javascript-3/hotel/6.jpg",
                },
              ]
            }
          />
        </ul>
      </section>
    </div>
  </main>
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="/"
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="img/logo.svg"
        width="64"
      />
    </a>
  </footer>
</div>
`;

exports[`FavoritePage render correctly FavoritePage without favorite offers 1`] = `
<div
  className="page page--favorites-empty"
>
  <Header>
    <User
      onLinkEmailClick={[Function]}
    />
  </Header>
  <main
    className="page__main page__main--favorites page__main--favorites-empty"
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites favorites--empty"
      >
        <h1
          className="visually-hidden"
        >
          Favorites (empty)
        </h1>
        <div
          className="favorites__status-wrapper"
        >
          <b
            className="favorites__status"
          >
            Nothing yet saved.
          </b>
          <p
            className="favorites__status-description"
          >
            Save properties to narrow down search or plan yor future trips.
          </p>
        </div>
      </section>
    </div>
  </main>
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="/"
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="img/logo.svg"
        width="64"
      />
    </a>
  </footer>
</div>
`;
